### ISSUE TITLE
1.7.1 Compilation errors on Solaris with SunStudio C++ 12.4

### ISSUE THREAD

**Issue by prrace** [harfbuzz/harfbuzz#630]
There were several issues here and I made source code changes .. probably
all of these could be wrapped in  #ifdef __SUNPRO_CC if so desired.

1)
harfbuzz/hb-face.cc", line 493: Error: The operation "extern "C" void(*)(void*) != void(*)(void*)" is illegal.

The line it did not like is
  if (face->destroy != _hb_face_for_data_closure_destroy)

Quite why that comparison is considered illegal is not apparent to me.
I don't know what the elegant solution is here. I simply hacked it to be
  if ((void*)(face->destroy) != (void*)_hb_face_for_data_closure_destroy)

and the compiler was happy, but I don't think of this as a long term solution.

2)
hb-dsalgs.hh", line 80: Error: Badly formed expression.
hb-dsalgs.hh", line 82: Error: Use ";" to terminate declarations.
hb-dsalgs.hh", line 82: Error: "," expected instead of ")".
hb-dsalgs.hh", line 82: Error: Use ";" to terminate declarations.
hb-dsalgs.hh", line 82: Error: A declaration was expected instead of ",".
hb-dsalgs.hh", line 83: Error: "," expected instead of ")".
hb-dsalgs.hh", line 85: Error: Use ";" to terminate declarations.
hb-dsalgs.hh", line 85: Error: A declaration was expected instead of ",".
hb-dsalgs.hh", line 85: Error: a is not defined.
hb-dsalgs.hh", line 85: Error: w is not defined.
hb-dsalgs.hh", line 86: Error: A declaration was expected instead of "if".
hb-dsalgs.hh", line 86: Error: a is not defined.

It turns out that the SunStudio compiler does not support "restrict" so this line was not parsed correctly :
static int sort_r_cmpswap(char *__restrict a, char *__restrict b, size_t w,

I changed it to
#ifndef __SUNPRO_CC
/* __restrict is same as restrict but better support on old machines */
static int sort_r_cmpswap(char *__restrict a, char *__restrict b, size_t w,
#else
static int sort_r_cmpswap(char *a, char *b, size_t w,
#endif

We lose the verification of __restrict but at least it compiles.

3) hb-string-array.hh", line 51: Error: An array cannot have zero size unless you use the option -features=zla.

That turns out to be the union member str in this code :
#undef _S
  } st;
  char str[0];

Since you need 'str' just to get the offset of the beginning of the array that is the other
member of this union and that is WAY bigger than the one extra byte I decided it was
easiest and harmless to just change it to

  char str[1];

**Comment by jfkthame** [comment]
> harfbuzz/hb-face.cc", line 493: Error: The operation "extern "C"
 > void(/)(void/) != void(/)(void/)" is illegal.
 >
 > The line it did not like is
 > if (face->destroy != _hb_face_for_data_closure_destroy)
 >
 > Quite why that comparison is considered illegal is not apparent to me.
 > I don't know what the elegant solution is here. I simply hacked it to be
 > if ((void*)(face->destroy) != (void*)_hb_face_for_data_closure_destroy)
 >
 > and the compiler was happy, but I don't think of this as a long term
 > solution.

Is it complaining because hb_destroy_func_t is declared (in hb-common.h) 
with extern "C" linkage, but _hb_face_for_data_closure_destroy is not; 
and so their types don't match?

If so, does it help if you wrap the _hb_face_for_data_closure_destroy 
function in hb-face.cc with extern "C" {...} as well?

**Comment by prrace** [comment]
In fact wrapping that declaration in extern "C" Is actually what I have now and I should
have said so in the first place.
The bug submission text was a copy/paste from an earlier email that is superseded.

**Comment by behdad** [comment]
> Is it complaining because hb_destroy_func_t is declared (in hb-common.h) with extern "C" linkage, but _hb_face_for_data_closure_destroy is not; and so their types don't match?

Right. But it's not clear why only this instance errs for Phil. (We sorted this out offline before; he's just reporting here now.)

**Comment by ebraminio** [comment]
We can not see these errors with ODS 12.6 https://circleci.com/gh/harfbuzz/harfbuzz/604

It is because of version difference or the fact our CI uses CMake? Can you help us on reproducing these issues by proposing needed change on these lines https://github.com/harfbuzz/harfbuzz/blob/master/.circleci/config.yml#L5-L13 on our CI?

**Comment by behdad** [comment]
Phil,

Does a cast help:
```
diff --git a/src/hb-face.cc b/src/hb-face.cc
index b8d7e84d..2fef09d0 100644
--- a/src/hb-face.cc
+++ b/src/hb-face.cc
@@ -514,7 +514,7 @@ hb_face_get_table_tags (hb_face_t    *face,
                        unsigned int *table_count, /* IN/OUT */
                        hb_tag_t     *table_tags /* OUT */)
 {
-  if (face->destroy != _hb_face_for_data_closure_destroy)
+  if (face->destroy != (hb_destroy_func_t) _hb_face_for_data_closure_destroy)
   {
     if (table_count)
       *table_count = 0;
```

**Comment by behdad** [comment]
> Is it complaining because hb_destroy_func_t is declared (in hb-common.h) with extern "C" linkage, but _hb_face_for_data_closure_destroy is not; and so their types don't match?
> 
> Right. But it's not clear why only this instance errs for Phil. (We sorted this out offline before; he's just reporting here now.)

Ok, I see why only this errs now. It's the comparison that is failing.  I suppose when trying to use _hb_face_for_data_closure_destroy as a destroy function the compiler does whatever magic cast to make it happen? So, I think an explicit cast might fix the comparison.
